### DISCUSSION TITLE
live:model in input not changing when inside a foreach and change happened from outside the for loop

### DISCUSSION THREAD

**Discussion by Ant-V** [livewire/livewire#7226] [1. Help]
I know the title is not the most descriptive, but I have no idea how to explain this in a sentence...

I have a `$fields` that is an array of `key:value` pairs. I need to have the user change them and add or remove as many as they want.

I have the following:
```` php
namespace App\Livewire;

use Livewire\Component;

class EditProjectFields extends Component
{
    public $fields = [];
    public function render()
    {
        return view('livewire.edit-project-fields');
    }
    public function addField()
    {
        $this->fields[] = ['key' => '', 'value' => ''];
    }
    public function removeField($index)
    {
        unset($this->fields[$index]);
        $this->fields = array_values($this->fields);
    }
}
````
```` html
<div>
    @foreach ($fields as $key => $value)
        <div class="field-pair" wire:key="field-{{ $key }}">
            <input type="text" wire:model.lazy="fields.{{ $key }}.key" placeholder="Key"/>
            {{$value['key']}}
            <input type="text" wire:model.lazy="fields.{{ $key }}.value" placeholder="Value"/>
            {{$value['value']}}
            <button wire:click="removeField({{ $key }})" type="button">Remove</button>
        </div>
    @endforeach

    <button wire:click="addField" type="button">Add Field</button>

    {{json_encode(($fields),JSON_PRETTY_PRINT)}}
</div>
````
Let's say I have this $fields: `[ { "key": "1", "value": "1" }, { "key": "2", "value": "2" }, { "key": "3", "value": "3" } ]` if I delete the second entry, the values outside the inputs change as they should, but the inputs show the numbers as if I deleted the last entry. If I go and change the values in the input, the correct value changes and checking the update response does indeed show the variable being updated correctly, but the values simply do not comply!
So the last pair would show as:
``` php
            <input type="text" wire:model.lazy="fields.{{ $key }}.key" placeholder="Key"/> <--2
            {{$value['key']}} <--3
```
Please help, I think I may be going insane...

**Comment by acharseth** [comment]
I think this is because you are "fooling" yourself when you name the index `$key` in the foreach loop. This `$key` is the index in the array, not the `key`value.
Debuging with `php -a` can be your friend in figuring this out. Eg:
```
php -a
php > $a=[ ["key" => "1", "value" => "1" ], [ "key" => "2", "value" => "2" ], [ "key" => "3", "value" => "3" ] ];
php > print_r ($a);
Array
(
    [0] => Array
        (
            [key] => 1
            [value] => 1
        )

    [1] => Array
        (
            [key] => 2
            [value] => 2
        )

    [2] => Array
        (
            [key] => 3
            [value] => 3
        )
)
php > unset($a[1]);
php > print_r($a);
Array
(
    [0] => Array
        (
            [key] => 1
            [value] => 1
        )

    [2] => Array
        (
            [key] => 3
            [value] => 3
        )

)
```
As you can see the array index has changed but not the `key` values.

**Reply by Ant-V** [reply]
Here is the thing though, `array_values` does change the keys

**Reply by acharseth** [reply]
No, not the values, only the array index is affected.
With my example above `print_r (array_values($a));` gives the same result as `print_r($a)` shown above where you can see that the values are not changed.
But may be I am misunderstanding you.

**Reply by Ant-V** [reply]
If you `unset($a[1]);` and `print_r($a);`, it should be different from `print_r (array_values($a));`, namely the keys are reordered.
The issue is that while it does change when I "echo" the array, the wire:model is still stuck on the wrong keys, as I'm trying to show in the last code block of the OG post

**Comment by joshhanley** [comment]
@Ant-V yeah @acharseth is correct, you if you use array indexes as keys, then Livewire won't actually know that it has changed, as before you had indexes 0,1,2, then remove 1 and reindex 2 to 1, so you are left with 0,1. Then Livewire thinks rows 0 and 1 are still the same and just removes 2, when it comes to `wire:model`, `wire:click` handlers, etc. but it will look correct as Livewire handles the contents of each element fine.

So best thing to do would be to not use the array index in your keys, instead your keys should be unique to the data in them.

You could add a unique ID to each of your fields by adding like an `id` key to each of your arrays, or instead of a numeric index, you could use a unique ID as your array index. Like in your `addField()` method:

```php
$this->fields[uniqid()] = ['key' => '', 'value' => ''];
```

Hope this helps!
